From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Otto=20Pfl=C3=BCger?= <otto.pflueger@abscue.de>
Date: Sat, 5 Apr 2025 16:06:55 +0200
Subject: [PATCH 019/225] iommu/sprd: fix handling of physical addresses
 outside 32-bit range
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Instead of saving the address to a 32-bit variable, shift it and keep
the page number in the variable.

Signed-off-by: Otto Pflüger <otto.pflueger@abscue.de>
---
 drivers/iommu/sprd-iommu.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/iommu/sprd-iommu.c b/drivers/iommu/sprd-iommu.c
index 4cd7bc2e6..1ca125339 100644
--- a/drivers/iommu/sprd-iommu.c
+++ b/drivers/iommu/sprd-iommu.c
@@ -319,7 +319,7 @@ static int sprd_iommu_map(struct iommu_domain *domain, unsigned long iova,
 	unsigned long flags;
 	unsigned int i;
 	u32 *pgt_base_iova;
-	u32 pabase = (u32)paddr;
+	u32 ppage = (u32)(paddr >> SPRD_IOMMU_PAGE_SHIFT);
 	unsigned long start = domain->geometry.aperture_start;
 	unsigned long end = domain->geometry.aperture_end;
 
@@ -338,8 +338,8 @@ static int sprd_iommu_map(struct iommu_domain *domain, unsigned long iova,
 
 	spin_lock_irqsave(&dom->pgtlock, flags);
 	for (i = 0; i < pgcount; i++) {
-		pgt_base_iova[i] = pabase >> SPRD_IOMMU_PAGE_SHIFT;
-		pabase += SPRD_IOMMU_PAGE_SIZE;
+		pgt_base_iova[i] = ppage;
+		ppage += 1;
 	}
 	spin_unlock_irqrestore(&dom->pgtlock, flags);
 
